README.md

```diff
@@ -901,6 +901,111 @@ _You can enable the following settings in Xcode by running [this script](resourc
   }
   ```
 
+* <a id='conditional-assignment'></a>(<a href='#conditional-assignment'>link</a>) **When initializing a new property with the result of a conditional statement (e.g. an if or switch statement), use a single if/switch expression where possible** rather than definining an uninitialized property and initializing it on every branch of the following conditional statement. [![SwiftFormat: conditionalAssignment](https://img.shields.io/badge/SwiftFormat-conditionalAssignment-7B0051.svg)](https://github.com/nicklockwood/SwiftFormat/blob/master/Rules.md#conditionalAssignment)
+
+  <details>
+
+  #### Why?
+
+  There are several benefits to using an if/switch expression over simply performing assignment on each branch of the following conditional statement:
+  1. In most cases, you no longer need to explicitly write a type annotation for the property that is being assigned to.
```

Comment:
```suggestion
  1. In most cases, you no longer need to explicitly write a type annotation for the variable that is being assigned to.
```

My understanding based on the code examples is that the rule only applies to variables. It's possible my terminology is off, though.

I proposed this edit so we can discuss whether "property" or "variable" (or both?) is most the most appropriate term.

Reply:
@calda thanks for confirming. I can see how a similar simplification could be made when implementing a property.

```swift
class Foo {
    let anotherNumber = 2
    
    // BEFORE

    var myValue: String {
        let result: String
        if anotherNumber % 2 == 0 {
            result = "even"
        } else {
            result = "odd"
        }
        return result
    }
    
    
    // RIGHT
    
    var myValue: String {
        if anotherNumber % 2 == 0 {
            "even"
        } else {
            "odd"
        }
    }
}
```

Would it make sense for this ^ similar autocorrection to be included in this rule? It's also possible that what I wrote above is distinct from the new Swift 5.9 functionality.

I tried to check if the above code would compile before Swift 5.9 but I am not sure if it's easy to do that now that I've downloaded Xcode 15. I updated my project to build for Swift 4.2 but even after doing that I added this function that uses the new Swift 5.9 conditional assignment feature and the application compiled 🤔 

<img width="723" alt="Screenshot 2023-11-08 at 8 40 30 AM" src="https://github.com/airbnb/swift/assets/1791049/6919eb1c-cad0-4c2f-96c8-c707b35ecbe3">


```swift
  func myFunc() {
    let value = if anotherNumber % 2 == 0 {
      true
    } else {
      false
    }
    print(value)
  }
```

Reply:
> ```swift
> // BEFORE
> var myValue: String {
>   let result: String
>   if anotherNumber % 2 == 0 {
>     result = "even"
>   } else {
>     result = "odd"
>   }
>   return result
> }
> 
> // AFTER
> var myValue: String {
>   if anotherNumber % 2 == 0 {
>     "even"
>   } else {
>     "odd"
>   }
> }
> ```
>
> Would it make sense for this ^ similar autocorrection to be included in this rule?
    
This is a transformation that could conceivably be implemented as a SwiftFormat rule, but currently isn't. As implemented the `conditionalAssignment` rule adopted in this proposal would convert the initial code to:

```swift
// BEFORE
var myValue: String {
  let result: String = 
    if anotherNumber % 2 == 0 {
      "even"
    } else {
      "odd"
    }
  return result
}
```

You could imagine a separate rule that would simplify away the `let result` / `return result` ceremony and just directly return the RHS expression. This could apply to any time where you define a variable and then return it immediately.

Another note is that if somebody writes this:

```swift
var myValue: String {
  if anotherNumber % 2 == 0 {
    return "even"
  } else {
    return "odd"
  }
}
```

we already automatically update it to:

```swift
var myValue: String {
  if anotherNumber % 2 == 0 {
    "even"
  } else {
    "odd"
  }
}
```

just by applying our existing "omit redundant return" rule. This happened automatically when we updated the SwiftFormat `--swiftversion` to 5.9. See https://github.com/airbnb/swift/pull/249 for a bit more discussion about this.

> I tried to check if the above code would compile before Swift 5.9 but I am not sure if it's easy to do that now that I've downloaded Xcode 15.

I use https://swiftfiddle.com/ to test older versions of the Swift compiler. It supports every version back to 2.2!

The ability to omit the `return` keyword in an if/switch statement like this is a new feature introduced in Swift 5.9.

> I updated my project to build for Swift 4.2 but even after doing that I added this function that uses the new Swift 5.9 conditional assignment feature and the application compiled 🤔

This "language version" setting is distinct from the "compiler version". The language version is a gating mechanism for breaking language changes. For example, our codebase currently has code that compiles in Swift 5.9 but won't compile in Swift 6.0. If you compile our app using the Swift 6.0 compiler, but with the "Swift 5 language mode", the Swift 6.0 compiler will permit all code that was valid under Swift 5. But since we'd be using the Swift 6.0 compiler, we'd also be able to use new features introduced in Swift 6.0.

Reply:
> > ```swift
> > // BEFORE
> > var myValue: String {
> >   let result: String
> >   if anotherNumber % 2 == 0 {
> >     result = "even"
> >   } else {
> >     result = "odd"
> >   }
> >   return result
> > }
> > 
> > // AFTER
> > var myValue: String {
> >   if anotherNumber % 2 == 0 {
> >     "even"
> >   } else {
> >     "odd"
> >   }
> > }
> > ```
> >
> > Would it make sense for this ^ similar autocorrection to be included in this rule?
> 
> This is a transformation that could conceivably be implemented as a SwiftFormat rule, but currently isn't. As implemented the `conditionalAssignment` rule adopted in this proposal would convert the initial code to:
> 
> ```swift
> // BEFORE
> var myValue: String {
>   let result: String = 
>     if anotherNumber % 2 == 0 {
>       "even"
>     } else {
>       "odd"
>     }
>   return result
> }
> ```
> 
> You could imagine a separate rule that would simplify away the `let result` / `return result` ceremony and just directly return the RHS expression. This could apply to any time where you define a variable and then return it immediately.

Makes sense! I agree that there's another rule that we could add in the future.

I wanted to discuss this now in case we thought that the best path would be expanding this currently proposed rule. Your observation that we could also condense property implementations that don't leverage conditional assignment, though, has convinced me that we would want a new rule.

> 
> Another note is that if somebody writes this:
> 
> ```swift
> var myValue: String {
>   if anotherNumber % 2 == 0 {
>     return "even"
>   } else {
>     return "odd"
>   }
> }
> ```
> 
> we already automatically update it to:
> 
> ```swift
> var myValue: String {
>   if anotherNumber % 2 == 0 {
>     "even"
>   } else {
>     "odd"
>   }
> }
> ```
> 
> just by applying our existing "omit redundant return" rule. This happened automatically when we updated the SwiftFormat `--swiftversion` to 5.9. See #249 for a bit more discussion about this.

Ah, thanks for mentioning this. I felt like there was a relationship between what I was writing about and omitting redundant return statements. You've made it clear to me!

> > I tried to check if the above code would compile before Swift 5.9 but I am not sure if it's easy to do that now that I've downloaded Xcode 15.
> 
> I use https://swiftfiddle.com/ to test older versions of the Swift compiler. It supports every version back to 2.2!

TIL about https://swiftfiddle.com/. So useful! I see that the simplified property implementation compiles in Swift 5.9 but not in Swift 5.8.

| Swift 5.8 | Swift 5.9 | 
| ---- | ---- |
| <img alt="Screenshot 2023-11-08 at 2 17 38 PM" src="https://github.com/airbnb/swift/assets/1791049/1ae207e0-9025-4eab-9694-dcea224e2716"> | <img alt="Screenshot 2023-11-08 at 2 17 24 PM" src="https://github.com/airbnb/swift/assets/1791049/4a574533-19d6-482f-9b60-77e0221ba80c"> |

> > I updated my project to build for Swift 4.2 but even after doing that I added this function that uses the new Swift 5.9 conditional assignment feature and the application compiled 🤔
> 
> This "language version" setting is distinct from the "compiler version". The language version is a gating mechanism for breaking language changes. For example, our codebase currently has code that compiles in Swift 5.9 but won't compile in Swift 6.0. If you compile our app using the Swift 6.0 compiler, but with the "Swift 5 language mode", the Swift 6.0 compiler will permit all code that was valid under Swift 5. But since we'd be using the Swift 6.0 compiler, we'd also be able to use new features introduced in Swift 6.0.

👍